src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp

```diff
@@ -105,7 +105,8 @@ inline oop ShenandoahBarrierSet::load_reference_barrier(oop obj, T* load_addr) {
   // Prevent resurrection of unreachable non-strorg references.
   if (!HasDecorator<decorators, ON_STRONG_OOP_REF>::value && obj != NULL &&
       _heap->is_concurrent_weak_root_in_progress() &&
-      !_heap->marking_context()->is_marked_strong(obj)) {
+      !(HasDecorator<decorators, ON_PHANTOM_OOP_REF>::value ? _heap->marking_context()->is_marked(obj)
```

Comment:
Maybe split out boolean locals for decorator tests? A matter of style, your call.

Reply:
Hmm yeah. Actually I am thinking how to do the HasDecorator tests as outermost tests, because those are the compile-time-tests. Would need to duplicate the whole test once for WEAK/UNKNOWN and once for PHANTOM I guess.